### DISCUSSION TITLE
CommandHandler - How to instantiate new aggregate with parameterized constructor

### DISCUSSION THREAD

**Discussion by dbrassard-genetec** [rasmus/EventFlow#901] [Q&A] [answered]
## Context
We have a Command that creates a new aggregate. 
When the CommandHandler executes the command, it creates a new aggregate instance using a default constructor.
We would like to create a new instance using a parameterized constructor so that we can enforce business rules in the constructor.

## Example
```csharp
public class CreateMyAggregateCommand : Command<MyAggregate, MyAggregateId, IExecutionResult>
{
        public CreateMyAggregateCommand (MyAggregateId id) : base(id)
        {
            
        }
}

public class CreateMyAggregateCommandHandler : CommandHandler<MyAggregate, MyAggregateId, IExecutionResult, CompleteProductUpdateCommand>
{
        public override Task<IExecutionResult> ExecuteCommandAsync(MyAggregate aggregate, CreateMyAggregateCommand command, CancellationToken cancellationToken)
        {
            // Here the aggregate instance is constructed using the default constructor.
            return Task.FromResult(executionResult);
        }
 }
```

## Questions
- Is it possible to have the command handler call a  parameterized constructor?
- Should we override the instance we a new instance?
- Is the aggregate creation a good use case for the command handler?

**Comment by rasmus** [comment, marked as answer]
Since you really don't know if the aggregate is already created at the time or it will be created by some competing thread I don't think doing it in the command handler is a good idea. Any information you would want to pass to the aggregate could as easily be passed via any methods and if you want some special factory for your aggregates, create a custom implementation for `IAggregateFactory`

**Reply by dbrassard-genetec** [reply]
Ok thank you for your answer.
